File: INTRODUCCION-POO/EJERCICIO-7/Celular.py
class Celular:
    def __init__(self, espacio_total_mb=1024):
        self.espacio_total_mb = espacio_total_mb
        self.espacio_ocupado_mb = 0
        self.aplicaciones_instaladas = []
        self.max_aplicaciones = 20
        self.bateria = 100

    def instalar_aplicacion(self, nombre, peso_mb):
        if len(self.aplicaciones_instaladas) < self.max_aplicaciones:
            if self.espacio_ocupado_mb + peso_mb <= self.espacio_total_mb:
                self.aplicaciones_instaladas.append({"nombre": nombre, "peso_mb": peso_mb})
                self.espacio_ocupado_mb += peso_mb
                print("-------------------------------------------------")
                print(f"Aplicación '{nombre}' instalada correctamente.")
                
            else:
                print(f"No hay suficiente espacio para instalar '{nombre}'.")
        else:
            print("Se ha alcanzado el límite de aplicaciones instaladas.")

    def utilizar_aplicacion(self, nombre_app, tiempo_uso_minutos):
        
        if self.bateria > 0:
            aplicacion = next((app for app in self.aplicaciones_instaladas if app["nombre"] == nombre_app), None)
            if aplicacion:
                uso_en_periodos = tiempo_uso_minutos // 10
                if uso_en_periodos > 0:
                    peso_mb = aplicacion["peso_mb"]
                    bateria_a_gastar = 0
                    if peso_mb > 250:
                        bateria_a_gastar = 5 * uso_en_periodos
                    elif peso_mb > 100:
                        bateria_a_gastar = 2 * uso_en_periodos
                    else:
                        bateria_a_gastar = 1 * uso_en_periodos

                    if self.bateria >= bateria_a_gastar:
                        self.bateria -= bateria_a_gastar
                        print(f"Utilizando '{nombre_app}' durante {tiempo_uso_minutos} minutos. Batería restante: {self.bateria}%.")
                    else:
                        print("Batería insuficiente para utilizar la aplicación.")
                        self.bateria = 0
                        print("Celular apagado.")
            else:
                print(f"La aplicación '{nombre_app}' no está instalada.")
        else:
            print("Celular apagado.")

    def mostrar_bateria(self):
        print(f"Batería restante: {self.bateria}%.")

# Crear un celular
mi_celular = Celular()

# Instalar algunas aplicaciones
mi_celular.instalar_aplicacion("WhatsApp", 80)
mi_celular.instalar_aplicacion("Instagram", 150)
mi_celular.instalar_aplicacion("JuegoPesado", 300)
mi_celular.instalar_aplicacion("TikTok", 90)
mi_celular.instalar_aplicacion("OtraApp", 50)
mi_celular.instalar_aplicacion("App1", 10)
mi_celular.instalar_aplicacion("App2", 10)
mi_celular.instalar_aplicacion("App3", 10)
mi_celular.instalar_aplicacion("App4", 10)
mi_celular.instalar_aplicacion("App5", 10)
mi_celular.instalar_aplicacion("App6", 10)
mi_celular.instalar_aplicacion("App7", 10)
mi_celular.instalar_aplicacion("App8", 10)
mi_celular.instalar_aplicacion("App9", 10)
mi_celular.instalar_aplicacion("App10", 10)
mi_celular.instalar_aplicacion("App11", 10)
mi_celular.instalar_aplicacion("App12", 10)
mi_celular.instalar_aplicacion("App13", 10)
mi_celular.instalar_aplicacion("App14", 10)
mi_celular.instalar_aplicacion("App15", 10) # Intentar instalar más de 20

# Mostrar batería inicial
print()
print()
mi_celular.mostrar_bateria()

# Utilizar algunas aplicaciones
mi_celular.utilizar_aplicacion("WhatsApp", 35)
mi_celular.utilizar_aplicacion("Instagram", 60)
mi_celular.utilizar_aplicacion("JuegoPesado", 20)
mi_celular.utilizar_aplicacion("TikTok", 90)

# Mostrar batería después de usar aplicaciones
mi_celular.mostrar_bateria()

# Utilizar una aplicación hasta que se acabe la batería
print("\nIntentando usar JuegoPesado hasta agotar la batería:")
while mi_celular.bateria > 0:
    mi_celular.utilizar_aplicacion("JuegoPesado", 10)

# Intentar usar una aplicación con la batería agotada
print("\nIntentando usar WhatsApp con la batería agotada:")
mi_celular.utilizar_aplicacion("WhatsApp", 15)

File: INTRODUCCION-POO/EJERCICIO-7/Celular.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Celular {
    private int espacioTotalMb;
    private int espacioOcupadoMb;
    private List<Map<String, Object>> aplicacionesInstaladas;
    private int maxAplicaciones;
    private int bateria;

    public Celular(int espacioTotalMb) {
        this.espacioTotalMb = espacioTotalMb;
        this.espacioOcupadoMb = 0;
        this.aplicacionesInstaladas = new ArrayList<>();
        this.maxAplicaciones = 20;
        this.bateria = 100;
    }

    public Celular() {
        this(1024);
    }

    public void instalarAplicacion(String nombre, int pesoMb) {
        if (aplicacionesInstaladas.size() < maxAplicaciones) {
            if (espacioOcupadoMb + pesoMb <= espacioTotalMb) {
                Map<String, Object> nuevaAplicacion = new HashMap<>();
                nuevaAplicacion.put("nombre", nombre);
                nuevaAplicacion.put("peso_mb", pesoMb);
                aplicacionesInstaladas.add(nuevaAplicacion);
                espacioOcupadoMb += pesoMb;
                System.out.println("-------------------------------------------------");
                System.out.println("Aplicacion '" + nombre + "' instalada correctamente.");

            } else {
                System.out.println("No hay suficiente espacio para instalar '" + nombre + "'.");
            }
        } else {
            System.out.println("Se ha alcanzado el limite de aplicaciones instaladas.");
        }
    }

    public void utilizarAplicacion(String nombreApp, int tiempoUsoMinutos) {
        if (bateria > 0) {
            Map<String, Object> aplicacion = null;
            for (Map<String, Object> app : aplicacionesInstaladas) {
                if (app.get("nombre").equals(nombreApp)) {
                    aplicacion = app;
                    break;
                }
            }

            if (aplicacion != null) {
                int usoEnPeriodos = tiempoUsoMinutos / 10;
                if (usoEnPeriodos > 0) {
                    int pesoMb = (int) aplicacion.get("peso_mb");
                    int bateriaAGastar = 0;
                    if (pesoMb > 250) {
                        bateriaAGastar = 5 * usoEnPeriodos;
                    } else if (pesoMb > 100) {
                        bateriaAGastar = 2 * usoEnPeriodos;
                    } else {
                        bateriaAGastar = 1 * usoEnPeriodos;
                    }

                    if (bateria >= bateriaAGastar) {
                        bateria -= bateriaAGastar;
                        System.out.println("Utilizando '" + nombreApp + "' durante " + tiempoUsoMinutos + " minutos. Bateria restante: " + bateria + "%.");
                    } else {
                        System.out.println("Bateria insuficiente para utilizar la aplicacion.");
                        bateria = 0;
                        System.out.println("Celular apagado.");
                    }
                }
            } else {
                System.out.println("La aplicacion '" + nombreApp + "' no esta instalada.");
            }
        } else {
            System.out.println("Celular apagado.");
        }
    }

    public void mostrarBateria() {
        System.out.println("Bateria restante: " + bateria + "%.");
    }

    public static void main(String[] args) {
        // Crear un celular
        Celular miCelular = new Celular();

        // Instalar algunas aplicaciones
        miCelular.instalarAplicacion("WhatsApp", 80);
        miCelular.instalarAplicacion("Instagram", 150);
        miCelular.instalarAplicacion("JuegoPesado", 300);
        miCelular.instalarAplicacion("TikTok", 90);
        miCelular.instalarAplicacion("OtraApp", 50);
        miCelular.instalarAplicacion("App1", 10);
        miCelular.instalarAplicacion("App2", 10);
        miCelular.instalarAplicacion("App3", 10);
        miCelular.instalarAplicacion("App4", 10);
        miCelular.instalarAplicacion("App5", 10);
        miCelular.instalarAplicacion("App6", 10);
        miCelular.instalarAplicacion("App7", 10);
        miCelular.instalarAplicacion("App8", 10);
        miCelular.instalarAplicacion("App9", 10);
        miCelular.instalarAplicacion("App10", 10);
        miCelular.instalarAplicacion("App11", 10);
        miCelular.instalarAplicacion("App12", 10);
        miCelular.instalarAplicacion("App13", 10);
        miCelular.instalarAplicacion("App14", 10);
        miCelular.instalarAplicacion("App15", 10); // Intentar instalar mas de 20

        // Mostrar bateria inicial
        System.out.println();
        System.out.println();
        miCelular.mostrarBateria();

        // Utilizar algunas aplicaciones
        miCelular.utilizarAplicacion("WhatsApp", 35);
        miCelular.utilizarAplicacion("Instagram", 60);
        miCelular.utilizarAplicacion("JuegoPesado", 20);
        miCelular.utilizarAplicacion("TikTok", 90);

        // Mostrar bateria despues de usar aplicaciones
        miCelular.mostrarBateria();

        // Utilizar una aplicacion hasta que se acabe la bateria
        System.out.println("\nIntentando usar JuegoPesado hasta agotar la bateria:");
        while (miCelular.bateria > 0) {
            miCelular.utilizarAplicacion("JuegoPesado", 10);
        }

        // Intentar usar una aplicacion con la bateria agotada
        System.out.println("\nIntentando usar WhatsApp con la bateria agotada:");
        miCelular.utilizarAplicacion("WhatsApp", 15);
    }
}
